File: rankboard-admin/server-python/app/services/rank_provider.py
"""RANK PROVIDER — the swappable-transport pattern, third appearance.

check_ranks() answers one question per keyword: "what position does
this project's domain hold on Google right now?" — via one of two
transports:

  DATAFORSEO_LOGIN/PASSWORD set  -> real lookups via DataForSEO's
                                    SERP API (Live mode: one POST,
                                    results in seconds)
  not set                        -> simulated random walk around the
                                    current rank, clearly labeled, so
                                    the feature is demoable for free

Callers (and the UI) only know "a lookup happened and here are the
numbers + the source". Swapping providers later (SerpApi, Search
Console) touches only this file.
"""
import base64
import concurrent.futures
import json
import random
import urllib.request

from ..config import (
    DATAFORSEO_BASE,
    DATAFORSEO_LOGIN,
    DATAFORSEO_PASSWORD,
    RANK_CHECK_DEPTH,
    RANK_LANGUAGE,
    RANK_LOCATION_CODE,
)


def check_ranks(domain: str | None, keywords: list[dict], location_code: int | None = None) -> tuple[dict, str]:
    """keywords: [{"term": str, "currentRank": int|None}, ...]
    location_code: the project's DataForSEO country code, or None to fall
    back to the global RANK_LOCATION_CODE default.
    Returns ({term: rank | None}, source). None = not found in the
    checked depth."""
    if DATAFORSEO_LOGIN and DATAFORSEO_PASSWORD:
        code = location_code if location_code is not None else RANK_LOCATION_CODE
        return _dataforseo(domain, [k["term"] for k in keywords], code), "dataforseo"
    return _simulated(keywords), "simulated"


def _simulated(keywords: list[dict]) -> dict:
    out = {}
    for k in keywords:
        cur = k.get("currentRank")
        base = cur if isinstance(cur, int) else random.randint(8, 45)
        out[k["term"]] = max(1, min(100, base + random.randint(-4, 3)))
    return out


def _domain_matches(item_domain: str | None, target: str) -> bool:
    d = (item_domain or "").lower()
    return d == target or d == f"www.{target}" or d.endswith("." + target)


def _dataforseo(domain: str, terms: list[str], location_code: int) -> dict:
    """DataForSEO's LIVE endpoint accepts only ONE task per request
    (batching returns 40000 "You can set only one task at a time"), so we
    send one POST per keyword. Those calls are independent, so we run them
    CONCURRENTLY — otherwise a 30-keyword project would take minutes.
    Each task asks Google for the top RANK_CHECK_DEPTH organic results for
    that keyword in the given country (location_code) / configured language."""
    auth = base64.b64encode(f"{DATAFORSEO_LOGIN}:{DATAFORSEO_PASSWORD}".encode()).decode()

    def _lookup(term: str):
        task = {
            "keyword": term,
            "location_code": location_code,
            "language_code": RANK_LANGUAGE,
            "device": "desktop",
            "depth": RANK_CHECK_DEPTH,
        }
        req = urllib.request.Request(
            f"{DATAFORSEO_BASE}/v3/serp/google/organic/live/advanced",
            data=json.dumps([task]).encode(),
            headers={"Authorization": f"Basic {auth}", "Content-Type": "application/json"},
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=60) as res:
            payload = json.loads(res.read())
        tasks = payload.get("tasks") or []
        if not tasks or tasks[0].get("status_code") != 20000:
            return None  # task failed; treat as "not found"
        for result in tasks[0].get("result") or []:
            for item in result.get("items") or []:
                # rank_group = position among ORGANIC results only;
                # rank_absolute would also count ads/SERP features.
                if item.get("type") == "organic" and _domain_matches(item.get("domain"), domain):
                    return item.get("rank_group")
        return None

    out: dict = {t: None for t in terms}
    if not terms:
        return out

    errors: list[Exception] = []
    with concurrent.futures.ThreadPoolExecutor(max_workers=min(8, len(terms))) as pool:
        future_to_term = {pool.submit(_lookup, t): t for t in terms}
        for future in concurrent.futures.as_completed(future_to_term):
            term = future_to_term[future]
            try:
                out[term] = future.result()
            except Exception as exc:  # network/HTTP failure for this one keyword
                errors.append(exc)

    # If EVERY request failed (bad auth, no balance, outage), surface it as
    # an error instead of silently reporting "all not found".
    if errors and len(errors) == len(terms):
        raise errors[0]
    return out

File: rankboard-admin/server-php/app/Services/RankProvider.php
<?php

namespace App\Services;

use Illuminate\Support\Facades\Http;
use RuntimeException;

/**
 * RANK PROVIDER — the swappable-transport pattern, third backend.
 *
 *   DATAFORSEO_LOGIN/PASSWORD set -> real lookups via DataForSEO's
 *                                    SERP API (Live mode)
 *   not set                       -> simulated random walk around the
 *                                    current rank, clearly labeled
 *
 * Callers only know "a lookup happened, here are the numbers + the
 * source". Swapping providers later touches only this file.
 */
class RankProvider
{
    /**
     * @param  array<int, array{term: string, currentRank: int|null}>  $keywords
     * @return array{0: array<string, int|null>, 1: string}  [ranks, source]
     */
    public static function checkRanks(?string $domain, array $keywords): array
    {
        $cfg = config('rankboard.dataforseo');
        if ($cfg['login'] && $cfg['password']) {
            return [self::dataForSeo($domain, array_column($keywords, 'term')), 'dataforseo'];
        }

        return [self::simulated($keywords), 'simulated'];
    }

    private static function simulated(array $keywords): array
    {
        $out = [];
        foreach ($keywords as $k) {
            $cur = $k['currentRank'];
            $base = is_int($cur) ? $cur : random_int(8, 45);
            $out[$k['term']] = max(1, min(100, $base + random_int(-4, 3)));
        }

        return $out;
    }

    private static function domainMatches(?string $itemDomain, string $target): bool
    {
        $d = strtolower($itemDomain ?? '');

        return $d === $target || $d === 'www.'.$target || str_ends_with($d, '.'.$target);
    }

    private static function dataForSeo(string $domain, array $terms): array
    {
        $rank = config('rankboard.rank');
        $cfg = config('rankboard.dataforseo');

        // One POST carries a task per keyword — the whole project is
        // checked in a single round trip.
        $tasks = array_map(fn (string $term) => [
            'keyword' => $term,
            'location_code' => $rank['location_code'],
            'language_code' => $rank['language'],
            'device' => 'desktop',
            'depth' => $rank['depth'],
        ], $terms);

        $response = Http::withBasicAuth($cfg['login'], $cfg['password'])
            ->timeout(60)
            ->post($cfg['base'].'/v3/serp/google/organic/live/advanced', $tasks);

        if (! $response->successful()) {
            throw new RuntimeException(
                'DataForSEO HTTP '.$response->status().': '.substr($response->body(), 0, 300)
            );
        }

        $payload = $response->json();
        $out = array_fill_keys($terms, null);

        // DataForSEO echoes data.keyword back lowercased & trimmed, so map
        // from that normalized form to our original term to look up $out —
        // matching on the raw string would drop any capitalized keyword.
        $byNormalized = [];
        foreach ($terms as $term) {
            $byNormalized[strtolower(trim($term))] = $term;
        }

        foreach (($payload['tasks'] ?? []) as $task) {
            $returned = $task['data']['keyword'] ?? null;
            $term = is_string($returned) ? ($byNormalized[strtolower(trim($returned))] ?? null) : null;
            if (($task['status_code'] ?? 0) !== 20000 || $term === null) {
                continue; // 20000 = this task succeeded
            }
            foreach (($task['result'] ?? []) as $result) {
                foreach (($result['items'] ?? []) as $item) {
                    // rank_group = position among ORGANIC results only;
                    // rank_absolute would also count ads/SERP features.
                    if (($item['type'] ?? '') === 'organic' && self::domainMatches($item['domain'] ?? null, $domain)) {
                        $out[$term] = $item['rank_group'] ?? null;
                        break;
                    }
                }
                if ($out[$term] !== null) {
                    break;
                }
            }
        }

        return $out;
    }
}
